2004-03-28  Jim Meyering  <jim@meyering.net>

	Minor optimization:
	* src/du.c (process_file): Don't record dev/inode for directories.

	Under some circumstances, without -c, du would mistakenly count the
	space of hard-linked files, not just the first one it encountered.
	Reported by Anthony Thyssen.
	* src/du.c (du_files): Don't ever clear the set of `seen' dev/inodes.

Index: src/du.c
===================================================================
RCS file: /cvsroot/coreutils/coreutils/src/du.c,v
retrieving revision 1.192
retrieving revision 1.193
diff -u -p -a -u -p -a -r1.192 -r1.193
--- src/du.c	27 Mar 2004 15:44:23 -0000	1.192
+++ src/du.c	28 Mar 2004 21:43:50 -0000	1.193
@@ -377,6 +377,7 @@ process_file (FTS *fts, FTSENT *ent)
      via a hard link, then don't let it contribute to the sums.  */
   if (skip
       || (!opt_count_all
+	  && ! S_ISDIR (sb->st_mode)
 	  && 1 < sb->st_nlink
 	  && hash_ins (sb->st_ino, sb->st_dev)))
     {
@@ -518,12 +519,6 @@ du_files (char **files, int bit_flags)
 	}
       FTS_CROSS_CHECK (fts);
 
-      /* This is a space optimization.  If we aren't printing totals,
-	 then it's ok to clear the duplicate-detection tables after
-	 each command line hierarchy has been processed.  */
-      if (ent->fts_level == 0 && ent->fts_info == FTS_D && !print_totals)
-	hash_clear (htab);
-
       process_file (fts, ent);
     }
 
